### ISSUE TITLE
Signing key rollover

### ISSUE THREAD

**Issue by Skoucail** [IdentityServer/IdentityServer4#864]
I'm trying to implement Signing key rollover.
To AddSigningCredential() i'm passing a certificate name from a certificate that is stored in my local machine. Now i would like to renew this certificate. Normal we delete the old expired certificate and import the new certificate. And i think this is why the IdentityServer has signing key rollover. 
But can't figure out how the AddValidationKeys() works.

- Should i keep the old certificate in the store and somehow get it from the certificate store by name and convert it to a AsymmetricSecurityKey?
- Or do i need to store the public certificate in my DB/Textfile/Certfile and try to convert it to a AsymmetricSecurityKey?

So basically what is the recommended way for signing key rollover in a real world example?

**Comment by brockallen** [comment]
Do the docs not help? https://identityserver4.readthedocs.io/en/release/topics/crypto.html#signing-key-rollover

**Comment by Skoucail** [comment]
@brockallen i did read the docs but i'm looking for a more technical explanation.
Where to store the old/new certificates (in a production environment), how to pass them to the AddValidationKeys function (as asymmetricsecuritykey is an abstract class).
I did look trough all the Identityserver4.sample projects but did not find any example where the AddValidationKeys function is used.

**Comment by brockallen** [comment]
> Where to store the old/new certificates (in a production environment).

It depends. On windows, use the certificate store, as this gives the easiest way to secure your keys. For the validation keys you don't need the private key, but you could still use the certificate store since the validation key is what's used to validate outstanding tokens. 

Maybe you could argue that somewhere else would be ok for validation keys (since no private key), but I could imagine a multi-faceted attack where the attacker can get IS to use their cert and then fool an app/API to accepting their token (might not be easy, but it's theoretical). My point is that just using loose x509 cer file from some random place on the file system might not the best. You'd have to think thru your infrastructure and think about all of these things. Maybe you are confident the filesystem is properly DACL'ed, then maybe it'd be ok.

**Comment by leastprivilege** [comment]
But to answer the technical question - load the cert into an instance of `X509SecurityKey`.

**Comment by Skoucail** [comment]
@brockallen @leastprivilege Thx for the help!
